common/main/java/com/couchbase/lite/AbstractDatabase.java

```diff
@@ -1180,36 +1180,51 @@ private void postDatabaseChanged() {
             if (!isOpen() || (c4DbObserver == null)) { return; }
 
             boolean external = false;
-            int nChanges;
             List<String> docIDs = new ArrayList<>();
-            do {
-                // Read changes in batches of MAX_CHANGES
-                final C4DatabaseChange[] c4DbChanges = c4DbObserver.getChanges(MAX_CHANGES);
-
-                int i = 0;
-                nChanges = (c4DbChanges == null) ? 0 : c4DbChanges.length;
-                if (nChanges > 0) {
-                    while (c4DbChanges[i] == null) {
-                        i++;
-                        nChanges--;
-                    }
+            while (true) {
+                final List<C4DatabaseChange> c4DbChanges = getDbChanges();
+                final int nChanges = c4DbChanges.size();
+
+                // if there are no more changes, we've gotten them all and we're done:
+                // post anything that hasn't been posted and get outta here.
+                if (nChanges <= 0) {
+                    postChanges(docIDs);
+                    return ;
                 }
-                final boolean newExternal = (nChanges > 0) && c4DbChanges[i].isExternal();
 
-                if ((!docIDs.isEmpty()) && ((nChanges <= 0) || (external != newExternal) || (docIDs.size() > 1000))) {
-                    dbChangeNotifier.postChange(new DatabaseChange((Database) this, docIDs));
+                // Core seems to promise that all changes in a batch
+                // will have the same external flag.  :shrug:
+                final boolean newExternal = c4DbChanges.get(0).isExternal();
+
+                if ((docIDs.size() > 1000) || (external != newExternal) ) {
```

Comment:
(external != newExternal && !docID.isEmpty()) so it doesn't need to do anything if docID is empty.

Reply:
Looks like this is handled inside of `postChanges`

Reply:
Yeah.... What Jim said.  To me it makes more sense that the thing that handles sending the updates is the thing that check to see if there are any.